kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/utils/HttpClientUtils.java

```diff
@@ -136,16 +142,26 @@ public static String basicCredentials(String username, String password) {
     return "Basic " + encoded;
   }
 
+  private static final AtomicBoolean WARNED = new AtomicBoolean();
+
   public static HttpClient createHttpClient(Config config) {
-    // TODO: replace with service load
-    try {
-        return ((HttpClient.Factory) Class.forName("io.fabric8.kubernetes.client.okhttp.OkHttpClientFactory")
-                .getDeclaredConstructor()
-                .newInstance()).createHttpClient(config);
-    } catch (InstantiationException | IllegalAccessException | IllegalArgumentException | InvocationTargetException
-            | NoSuchMethodException | SecurityException | ClassNotFoundException e) {
-        throw KubernetesClientException.launderThrowable(e);
+    ServiceLoader<HttpClient.Factory> loader = ServiceLoader.load(HttpClient.Factory.class);
+    HttpClient.Factory factory = null;
+    for (Iterator<HttpClient.Factory> iter = loader.iterator(); iter.hasNext();) {
+      HttpClient.Factory possible = iter.next();
+      if (factory != null && WARNED.compareAndSet(false, true)) {
+        LOGGER.warn("There are multiple httpclient implementation in the classpath, "
+            + "choosing the first non-okhttp implementation. "
+            + "You should exclude dependencies that aren't needed or use an explicit association of the HttpClient.Factory.");
+      }
+      if (factory == null || !possible.getClass().getName().equals("io.fabric8.kubernetes.client.okhttp.OkHttpClientFactory")) {
```

Comment:
As discussed during triage call, a less fragile approach to determine if this is the fallback/default client would be better.

Reply:
This has been updated to use isDefault, but I'm thinking we'll likely want to refine this again when we get around to #3887 - so that we can just have a single module with all client dependencies and system or config property to drive the selection.